### PR TITLE
adds new post about URLComponents

### PR DESCRIPTION
**Pull request by pbodsk**

### CODE REVIEW

**Comment by pbodsk** [comment]
Would love to hear you input on this one

File: source/_posts/2018-02-26-URLComponents.md [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+
+```swift
+func searchURL(with q: String, optionalParameters: [String: String]? = nil) -> URL? {
+    var urlString = "https://showsknownfrom.tv/search"
+
+    //Append query
+    urlString.append("?q=\(q)")
+
+    //Append parameters if any
+    if let optionalParameters = optionalParameters {
```
Comment (siemensikkema) [inline source/_posts/2018-02-26-URLComponents.md]: can be written as:
```swift
optionalParameters?.forEach { key, value in
    urlString.append("&\(key)=\(value)")
}
```

File: source/_posts/2018-02-26-URLComponents.md [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+And can have an optional URL parameter called `featured`. If featured is present and is true, we would like to show a flashing red `UILabel` with the caption "FEATURED" over our `WebView` (note: this feature has not been cleared with the designer yet!)
+
+## Composing and Decomposing URLs - The Hard Way
+### Composing URLs for Searching
+The stage is set and we are ready to look at how to call our search service in our pre-URLComponents world.
+
+To generate a `URL` in our iOS app, we could write some code like this:
+
+```swift
+func searchURL(with q: String, optionalParameters: [String: String]? = nil) -> URL? {
```
Comment (siemensikkema) [inline source/_posts/2018-02-26-URLComponents.md]: I'd prefer to have `parameters: [String: String] = [:]`

File: source/_posts/2018-02-26-URLComponents.md [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+    var urlString = "https://showsknownfrom.tv/search"
+
+    //Append query
+    urlString.append("?q=\(q)")
```
Comment (siemensikkema) [inline source/_posts/2018-02-26-URLComponents.md]: You could point out that URLComponents handle the case of not having query parameter nicely by not adding the `?` when there are none.

File: source/_posts/2018-02-26-URLComponents.md [hunk 4]
```diff
@@ -0,0 +1,367 @@
...
+func searchURL(with q: String, optionalParameters: [String: String]? = nil) -> URL? {
+    let urlString = "https://showsknownfrom.tv/search"
+
+    guard var urlComponents = URLComponents(string: urlString) else {
+        return nil
+    }
+
+    var queryItems: [URLQueryItem] = [URLQueryItem(name: "q", value: q)]
+
+    if let optionalParameters = optionalParameters {
```
Comment (siemensikkema) [inline source/_posts/2018-02-26-URLComponents.md]: You could do something similar as my suggestion is the non-URLQueryItems approach above

File: source/_posts/2018-02-26-URLComponents.md [hunk 5]
```diff
@@ -0,0 +1,367 @@
...
+
+Perfect!
+
+### Decomposing URLs
+Remember our example from before?
+
+Now that we know about `URLQueryItem`s, checking if a query parameter exists is soooo much easier:
+
+```swift
+func receivedUrl(_ url: URL, contains parameter: String) -> String? {
```
Comment (siemensikkema) [inline source/_posts/2018-02-26-URLComponents.md]: `receivedURL`

File: source/_posts/2018-02-26-URLComponents.md [hunk 6]
```diff
@@ -0,0 +1,367 @@
...
+    else {
+        return nil
+    }
+
+    let items = queryItems.filter({ $0.name == parameter })
+    return items.first?.value
+}
+
+if let url = URL(string: "https://showsknownfrom.tv/shows/12345678?featured=true") {
+  if let value = receivedUrl(url, contains: "featured") {
```
Comment (siemensikkema) [inline source/_posts/2018-02-26-URLComponents.md]: you could reduce this to 1 if statement with 2 `let`s

File: source/_posts/2018-02-26-URLComponents.md [hunk 7]
```diff
@@ -0,0 +1,367 @@
...
+Lets write an `extension` to `URL` so we can append query parameters and also ask if a URL contains a query parameter:
+
+```swift
+extension URL {
+  func append(queryParameters: [String: String]) -> URL? {
+      guard var urlComponents = URLComponents(url: self, resolvingAgainstBaseURL: true) else {
+          return nil
+      }
+
+      let urlQueryItems = queryParameters.map{
```
Comment (siemensikkema) [inline source/_posts/2018-02-26-URLComponents.md]: missing space after map - `map {`

File: source/_posts/2018-02-26-URLComponents.md [hunk 8]
```diff
@@ -0,0 +1,367 @@
...
+          return URLQueryItem(name: $0, value: $1)
+      }
+      urlComponents.queryItems = urlQueryItems
+      return urlComponents.url
+  }
+
+  func value(forParameter name: String) -> String? {
```
Comment (siemensikkema) [inline source/_posts/2018-02-26-URLComponents.md]: I normally prefer `valueForParameter(_ name:` or even `valueForParameter(withName name:`

File: source/_posts/2018-02-26-URLComponents.md [hunk 9]
```diff
@@ -0,0 +1,367 @@
...
+    guard let urlComponents = URLComponents(url: self, resolvingAgainstBaseURL: true),
+        let queryItems = urlComponents.queryItems else {
+            return nil
+    }
+    let items = queryItems.filter({ $0.name == name })
```
Comment (siemensikkema) [inline source/_posts/2018-02-26-URLComponents.md]: you can leave out the '(' and ')' or even write something like:
```
return URLComponents(url: self, resolvingAgainstBaseURL: true)?
    .queryItems?
    .filter { $0.name == name }?
    .first?
    .value
```

But that's a lot for one statement :P

File: source/_posts/2018-02-26-URLComponents.md [hunk 10]
```diff
@@ -0,0 +1,367 @@
...
+  }
+}
+```
+
+Poetry, no less!
+
+Lets test it out:
+
+```swift
+if let url = URL(string: "https://showsknownfrom.tv") {
```
Comment (siemensikkema) [inline source/_posts/2018-02-26-URLComponents.md]: 1 if statement?

File: source/_posts/2018-02-26-URLComponents.md [hunk 11]
```diff
@@ -0,0 +1,367 @@
...
+    if let appendedURL = url.append(queryParameters: ["q": "Halt and Catch Fire", "order" : "desc", "number": "100"]) {
+        print(appendedURL) //gives us: https://showsknownfrom.tv?q=Halt%20and%20Catch%20Fire&number=100&order=desc
+    }
+}
+
+if let url = URL(string: "https://showsknownfrom.tv/shows/12345678?featured=true&test=1") {
```
Comment (siemensikkema) [inline source/_posts/2018-02-26-URLComponents.md]: 1 if statement?

File: source/_posts/2018-02-26-URLComponents.md [hunk 12]
```diff
@@ -0,0 +1,367 @@
...
+
+`public func append(queryParameters: [String: String]) -> URL?`
+
+which will append a `String: String` dictionary to an already existing `URL` and return a new `URL` with the query parameters appended and paramter encoded.
+
+So, if you're into Cocoapods, add this to your `podfile`:
+
+`pod 'Codemine', '~>1.0.0'`
+
+And if you're more of a Carthage guy, add this to your `Cartfile`
```
Comment (siemensikkema) [inline source/_posts/2018-02-26-URLComponents.md]: "guy" -> "person"

**Comment by pbodsk** [comment]
Thanks for your comments @siemensikkema Good catches all around 👍 

I'll look into fixing them later today

File: source/_posts/2018-02-26-URLComponents.md (hunk 8 above)
Comment (pbodsk) [reply source/_posts/2018-02-26-URLComponents.md]: yeah I could be persuaded into the `valueForParameter(_ name:` variant maybe 😄 

I think the `valueForParameter(withName:` sounds ObjectiveC-ish (and there's nothing wrong with that).

The reason I choose the name is, that that is the name that was already in `Codemine` (I know, we could change it there too but thats the reason at least)

**Comment by pbodsk** [comment]
Integrated review comments from @siemensikkema 

Once again thanks 🙏

File: source/_posts/2018-02-26-URLComponents.md [hunk 13]
```diff
@@ -0,0 +1,363 @@
...
+//test
+if let url = URL(string: "https://showsknownfrom.tv/shows/12345678?featured=true"),
+   let parameter = receivedUrl(url, contains: "featured") {
+    print("found \(parameter)")
+} else {
+    print("not found")
+}
+```
+
+Pretty right? At least compared to the initial version.
```
Comment (roxanajula) [inline source/_posts/2018-02-26-URLComponents.md]: Pretty, right? - Missing a comma

**Pull request merged**